proxy/v1/config/cfg.proto

```
@@ -0,0 +1,413 @@
+  // labels, request attributes, etc.
+  HttpMatchCondition match = 1;
+
+  // TODO: add redirect support
+
+  // Each routing rule is associated with one or more upstream clusters,
+  // (see glossary in beginning of document). Weights associated with the
+  // cluster determine the proportion of traffic it receives.
+  repeated WeightedCluster weighted_clusters = 2;
+
+  // Faults can be injected into the API calls by the proxy, for
+  // testing the failure recovery capabilities of downstream services.
+  // Faults include aborting the Http request from downstream service,
+  // delaying the proxying of request to the upstream service, or both.
+  // TODO: This should move into UpstreamCluster
+  HttpFaultInjection fault = 3;
+
+  // Other things? such as tracing, etc. ?
+}
+
+// Basic routing rule match criterion using Tcp attributes and downstream
+// service tags
+message Layer4MatchCondition {
+  // Set of layer 4 match attributes such as src ip/port, dst ip/port and
+  // protocol
+  Layer4MatchAttributes layer4_attributes = 1;
+
+  // Identify the downstream cluster initiating the connection.
+  ClusterIdentifier src_cluster = 2;
+}
+
+// Layer4 connection match attributes
+message Layer4MatchAttributes {
+  // IPv4 or IPv6
+  string src_ip = 1;
+  // source port
+  uint32 src_port = 2;
+  // IPv4 or IPv6
+  string dst_ip = 3;
+  // destination port
+  uint32 dst_port = 4;
+  // protocol Tcp|Udp
+  string protocol = 5;
+}
+
+// One or more tags that uniquely identify an upstream or downstream
+// cluster. Tags can be arbitrary strings. The interpretation of tags is
+// specific to the platform (e.g., kubernetes vs VMs)
+message ClusterIdentifier {
+  repeated string tags = 1;
+}
+
+message WeightedCluster {
+  // Unique identity of the upstream cluster. Must be the same as those used
+  // in the definition of the UpstreamClusterPolicy.
+  ClusterIdentifier dst_cluster = 1;
+
+  // The proportion of connections to be forwarded to the upstream
```

On `Layer4MatchCondition`: As per @PiotrSikora 's comments, I have tried to avoid over-specializing this spec just for Envoy. Nginx and HAProxy support TCP/Udp proxing. We could add it to envoy over time.
I added fields to `Layer4MatchCondition` to make it the standard 5-tuple, where the protocol field can describe whether its UDP or TCP traffic. If its UDP and we are usign envoy, we could throw an error back to the user.